src/machine/machine-varlink.c

```diff
@@ -188,3 +188,90 @@ int vl_method_register(sd_varlink *link, sd_json_variant *parameters, sd_varlink
 
         return sd_varlink_reply(link, NULL);
 }
+
+static int lookup_machine_by_name(sd_varlink *link, Manager *manager, const char *machine_name, Machine **ret_machine) {
+        assert(link);
+        assert(manager);
+        assert(ret_machine);
+
+        if (!machine_name)
+                return -EINVAL;
+
+        if (!hostname_is_valid(machine_name, /* flags= */ VALID_HOSTNAME_DOT_HOST))
+                return -EINVAL;
+
+        Machine *machine = hashmap_get(manager->machines, machine_name);
+        if (!machine)
+                return -ESRCH;
+
+        *ret_machine = machine;
+        return 0;
+}
+
+static int lookup_machine_by_pid(sd_varlink *link, Manager *manager, pid_t pid, Machine **ret_machine) {
+        Machine *machine;
+        int r;
+
+        assert(link);
+        assert(manager);
+        assert(ret_machine);
+        assert_cc(sizeof(pid_t) == sizeof(uint32_t));
+
+        if (pid == 0) {
+                int pidfd = sd_varlink_get_peer_pidfd(link);
+                if (pidfd < 0)
+                        return log_debug_errno(pidfd, "Failed to get peer pidfd: %m");
+
+                r = pidfd_get_pid(pidfd, &pid);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to get pid from pidfd: %m");
+        }
+
+        if (pid <= 0)
+                return -EINVAL;
+
+        r = manager_get_machine_by_pid(manager, pid, &machine);
```

Comment:
hmm, we really should add manager_get_machine_by_pidref() btw, and always operate with PidRef structures. That's what we did in logind to track the session leader pid, and we should do it the same for the machine leader pid.

It's something we are doing across our codebase: use pidrefs whereevery we can.